strutil/limbuffer.go

```diff
@@ -0,0 +1,58 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2018 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package strutil
+
+import (
+	"bytes"
+)
+
+type LimitedBuffer struct {
+	buffer   *bytes.Buffer
+	maxLines int
+	maxBytes int
+}
+
+func NewLimitedBuffer(maxLines, maxBytes int) *LimitedBuffer {
+	return &LimitedBuffer{
+		buffer:   &bytes.Buffer{},
+		maxLines: maxLines,
+		maxBytes: maxBytes,
+	}
+}
+
+func (wr *LimitedBuffer) Write(data []byte) (int, error) {
+	sz, err := wr.buffer.Write(data)
+	if err != nil {
+		return 0, err
+	}
+	if wr.buffer.Len() > 2*wr.maxBytes {
+		data := wr.buffer.Bytes()
+		// don't limit by number of lines, we will apply line limit in Bytes(),
```

Comment:
I was a bit puzzled by this, maybe start the comment with something like: `Do an early (and inaccurate) truncate here to safe memory. The exact truncate happens in the .Bytes() function` (and then the rest of the comment)